server/src-lib/Hasura/RQL/Types/Function.hs

```diff
@@ -0,0 +1,62 @@
+module Hasura.RQL.Types.Function where
+
+import           Hasura.Prelude
+import           Hasura.RQL.Types.Common
+import           Hasura.SQL.Types
+
+import           Data.Aeson
+import           Data.Aeson.Casing
+import           Data.Aeson.TH
+import           Language.Haskell.TH.Syntax (Lift)
+
+import qualified Data.Sequence              as Seq
+import qualified Data.Text                  as T
+
+data FunctionType
+  = FTVOLATILE
+  | FTIMMUTABLE
+  | FTSTABLE
+  deriving (Eq)
+
+$(deriveJSON defaultOptions{constructorTagModifier = drop 2} ''FunctionType)
+
+funcTypToTxt :: FunctionType -> T.Text
+funcTypToTxt FTVOLATILE  = "VOLATILE"
+funcTypToTxt FTIMMUTABLE = "IMMUTABLE"
+funcTypToTxt FTSTABLE    = "STABLE"
+
+instance Show FunctionType where
+  show = T.unpack . funcTypToTxt
+
+newtype FunctionArgName =
+  FunctionArgName { getFuncArgNameTxt :: T.Text}
+  deriving (Show, Eq, ToJSON, FromJSON, Lift, DQuote, IsString)
+
+data FunctionArg
+  = FunctionArg
+  { faName       :: !(Maybe FunctionArgName)
+  , faType       :: !QualifiedPGType
+  , faHasDefault :: !Bool
+  } deriving (Show, Eq)
+
+$(deriveToJSON (aesonDrop 2 snakeCase) ''FunctionArg)
+
+data SessionArgument
+  = SessionArgument
+  { saName  :: !FunctionArgName
+  , saIndex :: !Int
+  } deriving (Show, Eq)
+$(deriveToJSON (aesonDrop 2 snakeCase) ''SessionArgument)
+
+data FunctionInfo
+  = FunctionInfo
+  { fiName          :: !QualifiedFunction
+  , fiSystemDefined :: !SystemDefined
+  , fiType          :: !FunctionType
+  , fiInputArgs     :: !(Seq.Seq FunctionArg)
+  , fiSessionVarArg :: !(Maybe SessionArgument)
```

Comment:
Keeping the session argument “on the side” like this seems like an overly indirect encoding to me, since it’s easy to just (accidentally) ignore it, and the code will still typecheck (but do the wrong thing). Ideally, when adding things like this, we should try to make it hard to ignore things that should not be ignored.

Generally, in Haskell, the simplest way to ensure something *cannot* be ignored is not to use a product type, but to use a sum. You can always just take one field out of a product type and ignore the rest, and often that happens unintentionally—you have code that uses a couple fields of a product, then you add a new field, and the old code still typechecks so you don’t think to update it—but you can’t do that with sums. Sums require you handle them with exhaustive case analysis, so when you add a new case, the compiler will tell you the places that need to be updated.

In this case, I think we can improve the representation in two ways with one change: we can make it hard to accidentally ignore the session argument, *and* we can avoid having to deal with explicit indexes (which are usually a pain in Haskell, since indexed datatypes are not inductive, and functional programming really likes inductive data structures). Here’s my proposal:

```haskell
newtype HasDefault = HasDefault { unHasDefault :: Bool }
  deriving (Show, Eq)

data FunctionArg
  = FAUserProvided !(Maybe FunctionArgName) !QualifiedPGType !HasDefault
  | FASessionVariables !FunctionArgName
  deriving (Show, Eq)

data FunctionInfo
  = FunctionInfo
  { ...
  , fiInputArgs :: !(Seq.Seq FunctionArg)
  , ...
  } deriving (Show, Eq)
```

This representation can drop the index because it’s just a property of an argument’s position in the sequence. It does technically allow `FASessionVariables` to appear more than once in the argument list, but that isn’t really a problem—we can handle it fine, even if we don’t actually provide a way for users to configure a function that way (since it wouldn’t be very useful), so it’s not as though we’ve made a truly illegal state representable.

Reply:
My initial type model is similar to what you mentioned above. But I removed it because
> It does technically allow FASessionVariables to appear more than once in the argument list

I want my type should convey that it has no session argument or only one session argument. Here's my proposal:
```haskell
newtype HasDefault = HasDefault { unHasDefault :: Bool }
  deriving (Show, Eq)

data FunctionArg
  = FunctionArg
  { faName       :: !(Maybe FunctionArgName)
  , faType       :: !QualifiedPGType
  , faHasDefault :: !HasDefault
  } deriving (Show, Eq)

newtype FunctionArgIndex = FunctionArgIndex {unFunctionArgIndex :: Int}
  deriving (Show, Eq)

data SessionArgument
  = SessionArgument
  { saName  :: !FunctionArgName
  , saIndex :: !FunctionArgIndex
  } deriving (Show, Eq)
$(deriveToJSON (aesonDrop 2 snakeCase) ''SessionArgument)

data InputArguments
  = IAWithSessionArgument !SessionArgument !(Seq.Seq FunctionArg)
  | IAWithoutSessionArgument !(Seq.Seq FunctionArg)
  deriving (Show, Eq)
```

Please give suggestions if any :slightly_smiling_face:

Reply:
I’m not very fond of dealing with indexes if it isn’t absolutely necessary. The reason is that indexes are a very indirect way to access data: if you have a list of elements and an index into that list, looking up the element is a partial operation. Put another way, representations that use indexes can represent a *lot* of invalid states, since most of the valid values of type `Int` do not correspond to any element of the list. Furthermore, if you manipulate the list to rearrange it in some way, you may have to keep the indexes in sync.

Sometimes indexes are necessary for some reason or another, but I don’t think they’re necessary here. To me, the illegal state of having multiple `FASessionVariables` values in the argument list is much less unpleasant than the illegal state of having an out-of-range index, since the former still has a totally reasonable interpretation for the downstream code and isn’t really “illegal” per se (it corresponds to an entirely valid function application), while the latter is unambiguously incorrect.

My other point of reasoning is that I was thinking about what Vamshi was talking about with respect to “argument presets,” which is a reasonable idea. Even if we don’t implement those right now, the approach I proposed models it well: it makes `FASessionVariables` into just a particular kind of argument preset. Would it ever be useful for a user to preset two different arguments of a function to be session variables? No, almost certainly not! But allowing it doesn’t really have any downside, either, which is why it doesn’t really bother me.

Reply:
Thank you @lexi-lambda for an exhaustive explanation.

> most of the valid values of type `Int` do not correspond to any element of the list

This point very convincing :). I've never thought in this way while using an `Int` before. I'll implement your proposal. However, the discussion is still open. We've to make use of the type system to represent our requirements correctly. `the types should speak` :). If we introduce "argument presets", then it demands significant refactor in types which also speaks about the change being made to the reviewer without being written explicitly in PR description.